src/MergeContext.js

```diff
@@ -454,6 +454,265 @@ class BranchPosition
+        return str.replace(/^\s*\n+/g, '').trimEnd();
+    }
+
+    // returns either a string containing canonical spelling of the first paragraph word
+    // that is followed by a colon, or null
+    _paragraphLabel(str) {
+        if (/^\s{4}/.test(str)) // not a regular paragraph but a quotation
+            return null;
+        const rawLabel = str.match(/^\s*(\S+)\s*:/);
+        if (!rawLabel)
+            return null;
+        const label = rawLabel[1].toLowerCase();
+        return label[0].toUpperCase() + label.substring(1);
+    }
+
+    // returns either paragraphLabel(x) if that value contains a dash, or null
+    _startsWithFieldName(str) {
+        const label = this._paragraphLabel(str);
+        return (label && label.includes('-')) ? label : null;
+    }
+
+    // basic checks for an unparsed message
+    _checkRaw(message) {
+        // CRs in CRLF sequences already removed
+        // other CRs are not treated specially (and are banned)
+        const lines = message.split('\n');
+        for (let i = 0; i < lines.length; ++i) {
+            const line = lines[i];
+            const invalidPosition = this._invalidCharacterPosition(line);
+
+            if (invalidPosition !== -1)
+                throw new Error(`bad character at line ${i}, offset ${invalidPosition}`);
+
+            // allow excessively long whitespace-only lines
+            // that some copy-pasted PR descriptions may include
+            const trimmedLine = this._trim(line);
+
+            // TODO: Allow longer header (and possibly even trailer) lines.
+            if (trimmedLine.length > 72)
+                throw new Error(`too long line '${trimmedLine}'`);
+        }
+    }
+
+    _parse(prDescriptionRaw) {
+        const prDescription = prDescriptionRaw.replace(/\r+\n/g, '\n');
+        this._checkRaw(prDescription);
+        const prDescriptionWithoutHeader = this._extractHeader(prDescription);
+        const prDescriptionWithoutHeaderAndTrailer = this._extractTrailer(prDescriptionWithoutHeader);
+        this._parseBody(prDescriptionWithoutHeaderAndTrailer);
+    }
+
+    // authorField is a {name, value, raw}
+    _parseAuthor(authorField) {
+        const trimmedValue = this._trim(authorField.value);
+        const cred = trimmedValue.match(/^([\w][^@<>,]*) <(\S+@\S+\.\S+)>$/);
+        if (!cred)
+            throw new Error(`unsupported ${authorField.name} value format: ${authorField.value}`);
+
+        if (cred[0].includes(','))
+            throw new Error(`${authorField.name} author name with a comma: ${authorField.value}`);
+
+        return {name: cred[1].trim(), email: cred[2].trim()};
+    }
+
+    _extractHeader(prDescriptionRaw) {
+        const prDescription = this._trim(prDescriptionRaw);
+        const headerFieldName = 'Authored-by';
+        if (this._startsWithFieldName(prDescription) === headerFieldName) {
+            let tokenizer = new FieldsTokenizer(prDescription, 1);
```

Comment:
FWIW, I find the old approach -- checking atEnd() rather than supplying the expected number of fields -- cleaner/simpler. I would also expect it to require fewer changes when we add support for more header fields. Perhaps I missed something -- I did not see why that approach needed to be replaced.

Reply:
I restored the old approach at cd5ad03.